### PR TITLE
Feature 406

### PR DESCRIPTION
**Pull request by sujata-m**
This PR contains following features - 

- Initial code for OSWValidation
- Added zipFileHandler class
- Added ExtractedDataValidator class
- Added unit test cases
- Updated README.md file

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+- It unzip the provided zip files
+- Check for nodes, edges and points geojson files inside the unzipped folder
+- Validate each file against schema, schema can be found here
+- Return true or false according to validation
+- you can check the error if it returned false.
+
+## Starting a new project with template
+
+- Add `python-lib-osw-validation` package as dependency in your `requirements.txt`
+- or `pip install python-lib-osw-validation`
```
Comment (susrisha) [inline README.md:22]: Does this work when the library is hosted in `testpy`?

File: src/python_lib_osw_validation/__init__.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+import jsonschema
+from typing import Dict, Any
+from .zipfile_handler import ZipFileHandler
+from .extracted_data_validator import ExtractedDataValidator
+
+SCHEMA_PATH = os.path.join(os.path.dirname(__file__), 'schema')
+
+
+class OSWValidation:
+    schema_dir = os.path.join(SCHEMA_PATH, 'opensidewalks.schema.json')
```
Comment (susrisha) [inline src/python_lib_osw_validation/__init__.py:12]: Please keep the schema file as input parameter. This allows us to do the validation on multiple schemas.

**Review by susrisha — CHANGES_REQUESTED** [review]
Two minor changes

File: README.md (hunk 1 above)
Comment (sujata-m) [reply README.md:22]: right now no, but once it is deployed to pypi then it would work

File: src/python_lib_osw_validation/__init__.py [hunk 3]
```diff
@@ -11,17 +11,27 @@
 class OSWValidation:
     schema_dir = os.path.join(SCHEMA_PATH, 'opensidewalks.schema.json')
```
Comment (susrisha) [inline src/python_lib_osw_validation/__init__.py:12]: probably rename this to `default_schema_file_path`

**Review by susrisha — CHANGES_REQUESTED** [review]
One more change requested

**Comment by sureshgaussian** [comment]
I get these message while running ./setup_env.sh. These may not be addressed as part of this PR itself, but consider adding a task to resolve these issues.

ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
spyder 5.3.3 requires pyqt5<5.16, which is not installed.
spyder 5.3.3 requires pyqtwebengine<5.16, which is not installed.
conda-repo-cli 1.0.20 requires clyent==1.2.1, but you have clyent 1.2.2 which is incompatible.
conda-repo-cli 1.0.20 requires nbformat==5.4.0, but you have nbformat 5.5.0 which is incompatible.

**Review by sureshgaussian — CHANGES_REQUESTED** [review]
Library doesnt expose errors to the users. Without this, the library is of limited use for the users. Please change the validator to expose the errors. See task 406

File: src/python_lib_osw_validation/__init__.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            data = json.load(file)
+        return data
+
+    def validate_osw_errors(self, geojson_data: Dict[str, Any]) -> bool:
+        '''Validate OSW Data against the schema and process all errors'''
+        validator = jsonschema.Draft7Validator(self.schema)
+        errors = validator.iter_errors(geojson_data)
+
+        error_count = 0
+        for error in errors:
```
Comment (sureshgaussian) [inline src/python_lib_osw_validation/__init__.py:85]: really poor coding here (I know this is exactly from the reference code, but we should rewrite it to make it better).
error_count = len(errors) will give you
and self.errors = errors is enough to expose errors out.

**Review by sureshgaussian — CHANGES_REQUESTED** [review]

File: src/python_lib_osw_validation/__init__.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+            with open(schema_path, 'r') as file:
+                schema = json.load(file)
+            return schema
+        except Exception as e:
+            self.error = e
+            raise Exception(f'Invalid or missing schema file: {e}')
+
+
+
+    def validate(self) -> bool:
```
Comment (susrisha) [inline src/python_lib_osw_validation/__init__.py:36]: redefine this method to return two variables, (isValid, errorList) . The list will have the errors and `isValid` will have the same bool as sent now.

**Review by susrisha — CHANGES_REQUESTED** [review]

File: setup_env.sh [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+env_name="tdei"
+# check if conda exists in the system
+conda_path=$(which conda)
+echo $conda_path
+echo "Checking if conda is installed on the system ..."
+if ! hash which conda 2>/dev/null
+then
+    echo " conda not found on the system"
+    echo " !!! Please install conda on your system before proceeding ..."
+    return 1
```
Comment (susrisha) [inline setup_env.sh:11]: replace with `exit 1`

**Review by sureshgaussian — CHANGES_REQUESTED** [review]
Running unittests resulted in failures.. is this expected?

python -m unittest discover -v tests/unit_tests 
test_invalid_empty_directory (test_extracted_data_validator.TestExtractedDataValidator) ... ok
test_invalid_missing_files_directory (test_extracted_data_validator.TestExtractedDataValidator) ... ok
test_invalid_missing_required_files_directory (test_extracted_data_validator.TestExtractedDataValidator) ... ok
test_valid_directory_structure (test_extracted_data_validator.TestExtractedDataValidator) ... ok
test_edges_invalid_zipfile (test_osw_validation.TestOSWValidation) ... ERROR
test_edges_invalid_zipfile_with_invalid_schema (test_osw_validation.TestOSWValidation) ... FAIL
test_edges_invalid_zipfile_with_schema (test_osw_validation.TestOSWValidation) ... ERROR
test_invalid_zipfile (test_osw_validation.TestOSWValidation) ... ERROR
test_invalid_zipfile_with_invalid_schema (test_osw_validation.TestOSWValidation) ... FAIL
test_invalid_zipfile_with_schema (test_osw_validation.TestOSWValidation) ... ERROR
test_missing_files_inside_zipfile (test_osw_validation.TestOSWValidation) ... ERROR
test_missing_files_inside_zipfile_with_invalid_schema (test_osw_validation.TestOSWValidation) ... FAIL
test_missing_files_inside_zipfile_with_schema (test_osw_validation.TestOSWValidation) ... ERROR
test_nodes_invalid_zipfile (test_osw_validation.TestOSWValidation) ... ERROR
test_nodes_invalid_zipfile_with_invalid_schema (test_osw_validation.TestOSWValidation) ... FAIL
test_nodes_invalid_zipfile_with_schema (test_osw_validation.TestOSWValidation) ... ERROR
test_points_invalid_zipfile (test_osw_validation.TestOSWValidation) ... ERROR
test_points_invalid_zipfile_with_invalid_schema (test_osw_validation.TestOSWValidation) ... FAIL
test_points_invalid_zipfile_with_schema (test_osw_validation.TestOSWValidation) ... ERROR
test_valid_zipfile (test_osw_validation.TestOSWValidation) ... ERROR
test_valid_zipfile_with_invalid_schema (test_osw_validation.TestOSWValidation) ... FAIL
test_valid_zipfile_with_schema (test_osw_validation.TestOSWValidation) ... ERROR
test_extract_invalid_zip (test_zipfile_handler.TestZipFileHandler) ... ok
test_extract_valid_zip (test_zipfile_handler.TestZipFileHandler) ... ok
test_remove_extracted_files (test_zipfile_handler.TestZipFileHandler) ... ok

======================================================================
ERROR: test_edges_invalid_zipfile (test_osw_validation.TestOSWValidation)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sureshdevalapalli/code/TDEI/TDEI-python-lib-osw-validation/tests/unit_tests/test_osw_validation.py", line 82, in test_edges_invalid_zipfile
    self.assertIsNotNone(result.error)
AttributeError: 'ValidationResult' object has no attribute 'error'

======================================================================
ERROR: test_edges_invalid_zipfile_with_schema (test_osw_validation.TestOSWValidation)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sureshdevalapalli/code/TDEI/TDEI-python-lib-osw-validation/tests/unit_tests/test_osw_validation.py", line 88, in test_edges_invalid_zipfile_with_schema
    self.assertIsNotNone(result.error)
AttributeError: 'ValidationResult' object has no attribute 'error'

======================================================================
ERROR: test_invalid_zipfile (test_osw_validation.TestOSWValidation)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sureshdevalapalli/code/TDEI/TDEI-python-lib-osw-validation/tests/unit_tests/test_osw_validation.py", line 46, in test_invalid_zipfile
    self.assertIsNotNone(result.error)
AttributeError: 'ValidationResult' object has no attribute 'error'

======================================================================
ERROR: test_invalid_zipfile_with_schema (test_osw_validation.TestOSWValidation)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sureshdevalapalli/code/TDEI/TDEI-python-lib-osw-validation/tests/unit_tests/test_osw_validation.py", line 52, in test_invalid_zipfile_with_schema
    self.assertIsNotNone(result.error)
AttributeError: 'ValidationResult' object has no attribute 'error'

======================================================================
ERROR: test_missing_files_inside_zipfile (test_osw_validation.TestOSWValidation)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sureshdevalapalli/code/TDEI/TDEI-python-lib-osw-validation/tests/unit_tests/test_osw_validation.py", line 118, in test_missing_files_inside_zipfile
    self.assertIsNotNone(result.error)
AttributeError: 'ValidationResult' object has no attribute 'error'

======================================================================
ERROR: test_missing_files_inside_zipfile_with_schema (test_osw_validation.TestOSWValidation)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sureshdevalapalli/code/TDEI/TDEI-python-lib-osw-validation/tests/unit_tests/test_osw_validation.py", line 124, in test_missing_files_inside_zipfile_with_schema
    self.assertIsNotNone(result.error)
AttributeError: 'ValidationResult' object has no attribute 'error'

======================================================================
ERROR: test_nodes_invalid_zipfile (test_osw_validation.TestOSWValidation)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sureshdevalapalli/code/TDEI/TDEI-python-lib-osw-validation/tests/unit_tests/test_osw_validation.py", line 64, in test_nodes_invalid_zipfile
    self.assertIsNotNone(result.error)
AttributeError: 'ValidationResult' object has no attribute 'error'

======================================================================
ERROR: test_nodes_invalid_zipfile_with_schema (test_osw_validation.TestOSWValidation)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sureshdevalapalli/code/TDEI/TDEI-python-lib-osw-validation/tests/unit_tests/test_osw_validation.py", line 70, in test_nodes_invalid_zipfile_with_schema
    self.assertIsNotNone(result.error)
AttributeError: 'ValidationResult' object has no attribute 'error'

======================================================================
ERROR: test_points_invalid_zipfile (test_osw_validation.TestOSWValidation)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sureshdevalapalli/code/TDEI/TDEI-python-lib-osw-validation/tests/unit_tests/test_osw_validation.py", line 100, in test_points_invalid_zipfile
    self.assertIsNotNone(result.error)
AttributeError: 'ValidationResult' object has no attribute 'error'

======================================================================
ERROR: test_points_invalid_zipfile_with_schema (test_osw_validation.TestOSWValidation)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sureshdevalapalli/code/TDEI/TDEI-python-lib-osw-validation/tests/unit_tests/test_osw_validation.py", line 106, in test_points_invalid_zipfile_with_schema
    self.assertIsNotNone(result.error)
AttributeError: 'ValidationResult' object has no attribute 'error'

======================================================================
ERROR: test_valid_zipfile (test_osw_validation.TestOSWValidation)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sureshdevalapalli/code/TDEI/TDEI-python-lib-osw-validation/tests/unit_tests/test_osw_validation.py", line 29, in test_valid_zipfile
    self.assertIsNone(result.error)
AttributeError: 'ValidationResult' object has no attribute 'error'

======================================================================
ERROR: test_valid_zipfile_with_schema (test_osw_validation.TestOSWValidation)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sureshdevalapalli/code/TDEI/TDEI-python-lib-osw-validation/tests/unit_tests/test_osw_validation.py", line 35, in test_valid_zipfile_with_schema
    self.assertIsNone(result.error)
AttributeError: 'ValidationResult' object has no attribute 'error'

======================================================================
FAIL: test_edges_invalid_zipfile_with_invalid_schema (test_osw_validation.TestOSWValidation)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sureshdevalapalli/code/TDEI/TDEI-python-lib-osw-validation/tests/unit_tests/test_osw_validation.py", line 91, in test_edges_invalid_zipfile_with_invalid_schema
    with self.assertRaises(Exception):
AssertionError: Exception not raised

======================================================================
FAIL: test_invalid_zipfile_with_invalid_schema (test_osw_validation.TestOSWValidation)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sureshdevalapalli/code/TDEI/TDEI-python-lib-osw-validation/tests/unit_tests/test_osw_validation.py", line 55, in test_invalid_zipfile_with_invalid_schema
    with self.assertRaises(Exception):
AssertionError: Exception not raised

======================================================================
FAIL: test_missing_files_inside_zipfile_with_invalid_schema (test_osw_validation.TestOSWValidation)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sureshdevalapalli/code/TDEI/TDEI-python-lib-osw-validation/tests/unit_tests/test_osw_validation.py", line 127, in test_missing_files_inside_zipfile_with_invalid_schema
    with self.assertRaises(Exception):
AssertionError: Exception not raised

======================================================================
FAIL: test_nodes_invalid_zipfile_with_invalid_schema (test_osw_validation.TestOSWValidation)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sureshdevalapalli/code/TDEI/TDEI-python-lib-osw-validation/tests/unit_tests/test_osw_validation.py", line 73, in test_nodes_invalid_zipfile_with_invalid_schema
    with self.assertRaises(Exception):
AssertionError: Exception not raised

======================================================================
FAIL: test_points_invalid_zipfile_with_invalid_schema (test_osw_validation.TestOSWValidation)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sureshdevalapalli/code/TDEI/TDEI-python-lib-osw-validation/tests/unit_tests/test_osw_validation.py", line 109, in test_points_invalid_zipfile_with_invalid_schema
    with self.assertRaises(Exception):
AssertionError: Exception not raised

======================================================================
FAIL: test_valid_zipfile_with_invalid_schema (test_osw_validation.TestOSWValidation)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sureshdevalapalli/code/TDEI/TDEI-python-lib-osw-validation/tests/unit_tests/test_osw_validation.py", line 38, in test_valid_zipfile_with_invalid_schema
    with self.assertRaises(Exception):
AssertionError: Exception not raised

----------------------------------------------------------------------
Ran 25 tests in 50.089s

FAILED (failures=6, errors=12)

**Review by sureshgaussian — APPROVED** [review]
I ran unit tests and example.py locally on my machine.

**Pull request merged**